core/executor/src/heap.rs

```diff
@@ -0,0 +1,303 @@
+				current_level += 1;
+				let has_buddy  = index & 1 == 1;
+				if has_buddy {
+					index += 1;
+					break 'up;
+				}
+			}
+		}
+
+		let current_level_offset = (1 << self.levels - current_level) - 1;
+		let level_offset = index - current_level_offset;
+		let block_offset = level_offset * (1 << current_level);
+
+		let ptr = BLOCK_SIZE as u32 * block_offset as u32;
+		self.allocated_bytes.insert(ptr, size as u32);
+
+		ptr
+	}
+
+	pub fn deallocate(&mut self, ptr_: u32) {
+		let ptr = &(ptr_ as u32);
+		let allocated_size = self.allocated_bytes.get(ptr).unwrap().clone() as u32;
+
+		let new_total_size = self.total_size - allocated_size;
+		if new_total_size < self.total_size {
+			trace!(target: "wasm-heap", "Heap size over {} Bytes after deallocation",  new_total_size);
+		}
+		self.total_size = new_total_size;
+
+		let count_blocks = (allocated_size as f32 / BLOCK_SIZE as f32).ceil() as u32;
+		let block_offset = ptr / BLOCK_SIZE as u32;
+		self.free(block_offset, count_blocks);
+		self.allocated_bytes.remove(ptr).unwrap();
+	}
+
+	pub fn free(&mut self, block_offset: u32, count_blocks: u32) {
+		let requested_level = Heap::get_tree_levels(count_blocks);
+		let current_level_offset = (1 << self.levels - requested_level) - 1;
+		let level_offset = block_offset / (1 << requested_level);
+		let index_offset = current_level_offset + level_offset;
+
+		if index_offset > self.tree.len() as u32 - 1 {
+			panic!("index offset {} is > length of tree {}", index_offset, self.tree.len());
+		}
+
+		self.free_and_merge(index_offset as usize);
+
+		let parent = self.get_parent_node_index(index_offset as usize);
+		self.update_parent_nodes(parent);
+	}
+
+	fn get_parent_node_index(&mut self, index: usize) -> usize {
+		(index + 1) / 2 - 1
+	}
+
+	fn free_and_merge(&mut self, index: usize) {
+		self.tree[index] = Node::Free;
+
+		if index == 0 {
+			return;
+		}
+
+		let other_node: usize;
+		let has_right_buddy = (index & 1) == 1;
+		if has_right_buddy {
+			other_node = index + 1;
+		} else {
+			other_node = index - 1;
+		}
+
+		if self.tree[other_node] == Node::Free {
+			let parent = self.get_parent_node_index(index);
+			self.free_and_merge(parent);
+		}
+	}
+
+	fn update_parent_nodes(&mut self, index: usize) {
+		let left_child = index * 2 + 1;
+		let right_child = index * 2 + 2;
+
+		let children_free = self.tree[left_child] == Node::Free && self.tree[right_child] == Node::Free;
+		let children_full = self.tree[left_child] == Node::Full && self.tree[right_child] == Node::Full;
+		if children_free {
+			self.tree[index] = Node::Free;
+		} else if children_full {
+			self.tree[index] = Node::Full;
+		} else {
+			self.tree[index] = Node::Split;
+		}
+
+		if index == 0 {
+			// Tree root
+			return;
+		}
+
+		let parent = self.get_parent_node_index(index);
+		self.update_parent_nodes(parent);
+	}
+
+	fn get_tree_levels(count_blocks: u32) -> u32 {
+		let requested_blocks;
+		if count_blocks == 0 {
+			requested_blocks = 1;
```

Comment:
I think the following *could* be faster:
```
if count_blocks == 0 {
    0
} else {
    let mut counter = 0;
    while count_blocks >>= 1 > 0 {
        counter += 1;
    }
    counter + 1
}
```
(This is the whole function)

Reply:
Thanks! Replaced the one in the code with this one slightly modified (because of an offset issue).